### PR TITLE
feat: add continuous integration github action

### PR DESCRIPTION
**Pull request by joey-kilgore**
# Overview
This script utilizes github actions to automatically test the `gnn_citation_networks.py` script with 2 different versions of python. It will run on any PR that merges into the main branch, as well as once a week ([see github actions docs for `cron` format](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#onschedule))

This will show automatically that the code either works or doesn't (including sending emails out to the repo admins). It will *ensure* that if code modifies any part of the scripts tested, then the final merged product will be fully functioning based on the script added in this PR. I would *strongly* recommend that the repo admins then change the branch rules of the repository to enforce a passing continuous integration check to allow for merges (and should enforce PR's with atleast one reviewer approval to merge into main)

For additional documentation on github actions see the docs [here](https://docs.github.com/en/actions)

# Discussion Points
I created this, being pretty new to this project, and in reference to #5 (which actually generates the same error there as well, [see the github action on my fork](https://github.com/joey-kilgore/sgnn-superneuro/actions/runs/14633763502/job/41060735751)) and so I don't know the following:
1. how long, assuming we fix the bug listed in #5, should this script take to run? It's probably not a good idea to have a github action linked to a job that takes 45min. Though it might be useful to have a job like that runs once a month just to ensure these long tasks keep functioning, but that kind of job doesn't need to run on every PR. I recommend that we choose a few setups. It is very easy to add Windows and MacOS runs as well, but those typically take a bit longer to run (due to machine allocation on the backend). Additionally, I have setup python3.9 and python3.12 support. I recommend that if there is cluster computer system that anyone in the group regularly uses and it has a specific version of python we use that (or if people are on newer/older versions than the ones listed here). And then we ensure the jobs we choose is <5min in length (note that it will run all OS+Python versions in parallel).
2. Is there another script we should run? I set it to the `gnn_citation_networks.py` because that's what is listed in the `README.md` but if there is another script that people suggest to be tested as well, then we should do that. Again, refer to (1) for notes on job length.

### CODE REVIEW

**Comment by kenblu24** [comment]
I've never done any CI before, and while I'd be excited to try it, I'm not sure it makes sense for this repository.

The central question I have is Who is this for?

1. The script mentioned is a training script. So are you recommending that getting the results of training is a regularly scheduled thing? Most of the people on this repo just run the script on their machine or whatever workstation or cluster they have access to. This is also hacky research code, not meant to be used in any production environment, so I don't think it matters if the code is continually tested for functionality. It just needs to work when we want the result for the paper.

2. There are several branches on this repo where each student or contributor is working on their own thing, i.e. long-lived branches for each person.
  a. Is the benefit that others can easily see whether the code is in a working state or not? We already have weekly meetings to discuss and assist each other on what we're working on. Does each person have to instead maintain the CI for their own workflow and environment, and would this be a burden?

I can see how CI would be useful for tools but I fail to see what the benefit is for this specific repo given the engineering cost. That said, this might be the push I need to finally learn how to do CI on my stuff.

**Comment by kenblu24** [comment]
#5 would not have been caught by CI unless we manually run it or connect it to commits on SuperNeuroMAT, as the breaking change there happened after the latest commit to `main` on this repo.

**Comment by joey-kilgore** [comment]
These are some good questions.
**Who is this for -** Potentially other people like me, if they are sent this project as their starting point, it may be helpful to have some piece of code that is in the readme and functioning for them to get this working. Maybe I am the only person this will happen for, it seems like this repository is more of a test-bench area for testing as opposed to a published repo meant for others to install and tinker with. So, as you point out, this might make the entire question of CI moot.

**Recommending getting the results of training as a regularly scheduled thing -** Not necessarily. The entire point of CI is purely to ensure that the code is functioning. That might mean you do extremely short training, on a small dataset, and see little improvement in accuracy, but you will at least know that the code is functioning and not throwing errors. It seems given the level of flux this repo is regularly under, that this might not be helpful, but it would help anyone who is new and enters this repository (see above point that this might only be me)

**would this be a burden -** Everyone could continue with what they are doing. The only thing this would enforce is that the script can run *and it only checks it when you open a PR to merge with the main branch*. It will never come into your workflow on your personal branch until you decide it should be merged into main. If the repo is under so much change that this likely isn't the case, and it is unimportant for this script to keep running then this isn't needed.

**unless we manually run it or connect it to SuperNeuroMAT - ** Actually, the solution I provided here also runs the script weekly. This means that while it won't catch it immediately, it would have caught this issue automatically (which may or may not be helpful).

It seems this repository is much more of a test-bench than the initial impression I was given, so I am not sure this is useful/wanted for this environment. If other's feel differently they can comment and reopen this PR.

**Comment by kenblu24** [comment]
Doing a short, small training run might be useful for SuperNeuroMAT itself. I did some googling and it seems like it's possible to run actions triggered by events on another repo; this might allow us to check if changes in SuperNeuroMAT breaks existing scripts like we have here. We could isolate this to a special testing branch on this repo to avoid tons of actions running. It's a little complicated to set this up though. @prasannadate Putting this on your radar
